store/iterator.go

```diff
@@ -108,163 +95,158 @@ func (b *btreeIter) wrap(parent Iterator) *itemIter {
-		return i.wrap.get().Key()
-	case parent:
-		return i.parent.Key()
-	default: //none
-		panic("Advanced past the end!")
+// advance will read next from wrap iterators,
+// and set cached value as well as done flags.
+//
+// It will skip any deleted items before the i.cachedParent.Key value
+//
+// it will skip closed and missing iterators.
+// doesn't return ErrDone, but only unexpected data errors.
+func (i *itemIter) advanceWrap() error {
+	for !i.wrapDone && i.cachedWrap == nil {
+		var err error
+		i.cachedWrap, err = i.wrap.Next()
+		if errors.ErrDone.Is(err) {
+			i.wrapDone = true
+		} else if err != nil {
+			return errors.Wrap(err, "advance wrap")
+		} else if _, ok := i.cachedWrap.(deletedItem); ok {
+			// if delete, stop if we are at or higher than the parent key
+			if i.cachedParent.Key != nil && bytes.Compare(i.cachedWrap.Key(), i.cachedParent.Key) >= 0 {
+				break
+			}
+			// otherwise, unset cachedWrap, so we read more
+			i.cachedWrap = nil
+		}
+		// if it is a setItem, we break out
 	}
+	return nil
 }
 
-// Value returns the value of the cursor.
-func (i *itemIter) Value() (value []byte) {
-	switch i.firstKey() {
-	case us, both:
-		return i.wrap.get().(setItem).value
-	case parent:
-		return i.parent.Value()
-	default: // none
-		panic("Advanced past the end!")
+// Next moves the iterator to the next sequential key in the database, as
+// defined by order of iteration.
+//
+// If Valid returns false, this method will panic.
+func (i *itemIter) Next() (key, value []byte, err error) {
+	// this guarantees that both have xxxDone == true or cachedXxx != nil
+	if err := i.advanceParent(); err != nil {
+		return nil, nil, errors.Wrap(err, "advanceParent")
+	}
+	// advances the wrap and skips all deleted up to parent key
+	if err := i.advanceWrap(); err != nil {
+		return nil, nil, errors.Wrap(err, "advanceWrap")
 	}
-}
-
-// Close releases the Iterator.
-func (i *itemIter) Close() {
-	i.parent.Close()
-	i.wrap.close()
-}
 
-// skipAllDeleted loops and skips any number of deleted items
-func (i *itemIter) skipAllDeleted() error {
-	var err error
-	more := true
-	for more {
-		more, err = i.skipDeleted()
-		if err != nil {
-			return err
-		}
+	if i.wrapDone {
+		return i.returnCachedParent()
+	}
+	if i.parentDone {
+		return i.returnCachedWrap()
 	}
-	return nil
-}
 
-// skipDeleted jumps over all elements we can safely fast forward
-// return true if skipped, so we can skip again
-func (i *itemIter) skipDeleted() (bool, error) {
-	src := i.firstKey()
-	if src == us || src == both {
-		// if our next is deleted, advance...
-		if _, ok := i.wrap.get().(deletedItem); ok {
-			i.wrap.next()
-			// if parent had the same key, advance parent as well
-			if src == both {
-				err := i.parent.Next()
-				if err != nil {
-					return false, err
-				}
-			}
-			return true, nil
+	// both are valid... try it
+	switch bytes.Compare(i.cachedParent.Key, i.cachedWrap.Key()) {
+	case 1: // cachedWrap lower
+		i.returnCachedWrap()
```

Comment:
req: you are not returning anything here and in L 212 so that this will cause the panic in L224

Reply:
totally right, I need to return there.
I can't believe the tests passed